File: kvdom/kvdom-dom/src/wasm32Main/resources/Node.js
konan.libraries.push({
    kvdom_Node_appendChild: function (arena, obj, childArena, childObj) {
        const parent = kotlinObject(arena, obj);
        const child = kotlinObject(childArena, childObj);
        parent.appendChild(child);
    },
    kvdom_Node_removeChild: function (arena, obj, childArena, childObj) {
        const parent = kotlinObject(arena, obj);
        const child = kotlinObject(childArena, childObj);
        parent.removeChild(child);
    },
    kvdom_Node_hasChildNodes: function (arena, obj) {
        const bool = kotlinObject(arena, obj).hasChildNodes();
        return bool === undefined ? -1 : bool ? 1 : 0;
    },
    kvdom_Node_childNodes: function (arena, obj) {
        const nodes = kotlinObject(arena, obj).childNodes;
        return toArena(arena, nodes);
    },
});

File: kvdom/kvdom-dom/src/wasm32Main/kotlin/lt/petuska/kvdom/dom/Node.kt
package lt.petuska.kvdom.dom

import kotlinx.wasm.jsinterop.*
import lt.petuska.kvdom.dom.util.*

public actual abstract class Node(arena: Arena, index: Object) : EventTarget(arena, index) {
  public actual open var textContent: String?
    get() = getStringProperty("textContent")
    set(value) {
      setStringProperty("textContent", value)
    }

  public actual fun removeChild(child: Node): Node {
    js_removeChild(arena, index, child.arena, child.index)
    return child
  }

  public actual fun appendChild(node: Node): Node {
    js_appendChild(arena, index, node.arena, node.index)
    return node
  }

  public actual val parentNode: Node?
    get() = getObjProperty("parentNode")?.let {
      object : Node(it.arena, it.index) {}
    }

  public actual val firstChild: Node?
    get() = getObjProperty("firstChild")?.let {
      object : Node(it.arena, it.index) {}
    }

  public actual fun hasChildNodes(): Boolean = js_hasChildNodes(arena, index) == 1
}

@SymbolName("kvdom_Node_hasChildNodes")
private external fun js_hasChildNodes(
  arena: Arena,
  index: Object,
): Int

@SymbolName("kvdom_Node_appendChild")
private external fun js_appendChild(
  arena: Arena,
  index: Object,
  childArena: Arena,
  childIndex: Object,
)

@SymbolName("kvdom_Node_removeChild")
private external fun js_removeChild(
  arena: Arena,
  index: Object,
  childArena: Arena,
  childIndex: Object,
)
